sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -988,7 +989,15 @@ private[hive] class HiveClientImpl(
 private[hive] object HiveClientImpl extends Logging {
   /** Converts the native StructField to Hive's FieldSchema. */
   def toHiveColumn(c: StructField): FieldSchema = {
-    val typeString = HiveVoidType.replaceVoidType(c.dataType).catalogString
+    val typeString = if (c.metadata.contains(CHAR_VARCHAR_TYPE_STRING_METADATA_KEY)) {
+      // For Hive Serde, we still need to to restore the raw type for char and varchar type.
+      // When reading data in parquet, orc, or avro file format with string type for char,
+      // the tailing spaces may lost if we are not going to pad it.
+      c.metadata.getString(CHAR_VARCHAR_TYPE_STRING_METADATA_KEY)
```

Comment:
we can add a `getRawTypeString` in `CharVarcharUtils`, instead of exposing `CHAR_VARCHAR_TYPE_STRING_METADATA_KEY`?